File: app/src/main/java/com/sami/rippel/utils/BitmapUtils.java
package com.sami.rippel.utils;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.ColorFilter;
import android.graphics.LightingColorFilter;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;

import com.sami.rippel.model.ViewModel;

import java.io.File;

/**
 * Created by souna on 12/17/2017.
 */

public class BitmapUtils {

    public static Bitmap rotateBitmap(Bitmap source, float angle) {
        if (!source.isRecycled()) {
            Matrix matrix = new Matrix();
            matrix.postRotate(angle);
            return Bitmap.createBitmap(source, 0, 0, source.getWidth(),
                    source.getHeight(), matrix, true);
        } else
            return null;
    }

    public Bitmap setReducedImageSize(String imagePath) {

        int targetImageViewWidth = ViewModel.Current.device.getScreenHeightPixels();
        int targetImageViewHeight = ViewModel.Current.device.getScreenWidthPixels();
        BitmapFactory.Options bmOptions = new BitmapFactory.Options();
        bmOptions.inJustDecodeBounds = true;
        BitmapFactory.decodeFile(imagePath, bmOptions);
        int cameraImageWidth = bmOptions.outWidth;
        int cameraImageHeight = bmOptions.outHeight;
        bmOptions.inSampleSize = Math.min(cameraImageWidth / targetImageViewWidth, cameraImageHeight / targetImageViewHeight);
        bmOptions.inJustDecodeBounds = false;
        return BitmapFactory.decodeFile(imagePath, bmOptions);
    }


    static Bitmap decodeSampledBitmapFromFile(String path, int reqWidth,
                                              int reqHeight) {
        final BitmapFactory.Options options = new BitmapFactory.Options();
        options.inJustDecodeBounds = true;
        BitmapFactory.decodeFile(path, options);
        final int height = options.outHeight;
        final int width = options.outWidth;
        options.inPreferredConfig = Bitmap.Config.ARGB_8888;
        int inSampleSize = 1;
        if (height > reqHeight) {
            inSampleSize = Math.round((float) height / (float) reqHeight);
        }
        int expectedWidth = width / inSampleSize;
        if (expectedWidth > reqWidth) {
            inSampleSize = Math.round((float) width / (float) reqWidth);
        }
        options.inSampleSize = inSampleSize;
        options.inJustDecodeBounds = false;
        return BitmapFactory.decodeFile(path, options);
    }

    public Bitmap getPreview(String fileName) {
        File image = new File(fileName);
        BitmapFactory.Options bounds = new BitmapFactory.Options();
        bounds.inJustDecodeBounds = true;
        BitmapFactory.decodeFile(image.getPath(), bounds);
        if ((bounds.outWidth == -1) || (bounds.outHeight == -1)) {
            return null;
        }
        //int originalSize = (bounds.outHeight > bounds.outWidth) ? bounds.outHeight
        //        : bounds.outWidth;
        //BitmapFactory.Options opts = new BitmapFactory.Options();
        //opts.inSampleSize = originalSize / 20;
        return BitmapFactory.decodeFile(image.getPath());
    }

    public Bitmap convertDrawableToBitmap(Drawable drawable) {
        if (drawable instanceof BitmapDrawable) {
            return ((BitmapDrawable) drawable).getBitmap();
        }

        Bitmap bitmap = Bitmap.createBitmap(drawable.getIntrinsicWidth(),
                drawable.getIntrinsicHeight(), Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(bitmap);
        drawable.setBounds(0, 0, canvas.getWidth(), canvas.getHeight());
        drawable.draw(canvas);

        return bitmap;
    }


    public Drawable covertBitmapToDrawable(Context context, Bitmap bitmap) {
        return new BitmapDrawable(context.getResources(), bitmap);
    }


    public Bitmap changeImageColor(Bitmap sourceBitmap, int color) {
        try {
            Bitmap resultBitmap = Bitmap.createBitmap(sourceBitmap, 0, 0,
                    sourceBitmap.getWidth() - 1, sourceBitmap.getHeight() - 1);
            Paint p = new Paint();
            ColorFilter filter = new LightingColorFilter(color, 1);
            p.setColorFilter(filter);
            Canvas canvas = new Canvas(resultBitmap);
            canvas.drawBitmap(resultBitmap, 0, 0, p);
            return resultBitmap;
        } catch (Exception e) {
            return null;
        }
    }


}

File: base-android/src/main/java/com/sfaxdroid/base/utils/BitmapUtils.kt
package com.sfaxdroid.base.utils

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.LightingColorFilter
import android.graphics.Paint
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable

/**
 * Created by souna on 12/17/2017.
 */
object BitmapUtils {

    fun changeImageColor(
        sourceBitmap: Bitmap,
        color: Int,
        width: Int = 0,
        height: Int = 0
    ): Bitmap? {
        return try {
            val resultBitmap: Bitmap = if (width != 0) {
                Bitmap.createScaledBitmap(
                    sourceBitmap, width, height,
                    false
                )
            } else {
                Bitmap.createBitmap(
                    sourceBitmap, 0, 0,
                    sourceBitmap.width - 1, sourceBitmap.height - 1
                )
            }
            val canvas = Canvas(resultBitmap)
            canvas.drawBitmap(resultBitmap, 0f, 0f, Paint().apply {
                colorFilter = LightingColorFilter(color, 1)
            })
            resultBitmap
        } catch (e: Exception) {
            null
        }
    }

    private fun covertBitmapToDrawable(context: Context, bitmap: Bitmap?): Drawable {
        return BitmapDrawable(context.resources, bitmap)
    }

    fun covertToDrawableWithColorChange(
        context: Context,
        selectedColor: Int,
        drawable: Drawable
    ): Drawable {
        return covertBitmapToDrawable(
            context,
            changeImageColor(
                convertDrawableToBitmap(
                    drawable
                ), selectedColor
            )
        )
    }

    private fun convertDrawableToBitmap(drawable: Drawable): Bitmap {
        if (drawable is BitmapDrawable) {
            return drawable.bitmap
        }
        val bitmap = Bitmap.createBitmap(
            drawable.intrinsicWidth,
            drawable.intrinsicHeight, Bitmap.Config.ARGB_8888
        )
        val canvas = Canvas(bitmap)
        drawable.setBounds(0, 0, canvas.width, canvas.height)
        drawable.draw(canvas)
        return bitmap
    }
}
